### DISCUSSION TITLE
Required/NotRequired and inheritance

### DISCUSSION THREAD

**Discussion by JelleZijlstra** [python/typing#1516] [General]
Consider this code:

```python
from typing import NotRequired, Required, TypedDict

class A(TypedDict):
    a: NotRequired[int]
    b: Required[int]

class B(A):
    a: Required[int]
    b: NotRequired[int]
```

[Mypy](https://mypy-play.net/?mypy=latest&python=3.11&gist=0c198703a5d3d9c8eb5befe88bc3ba20) says:
```
main.py:8: error: Overwriting TypedDict field "a" while extending  [misc]
main.py:9: error: Overwriting TypedDict field "b" while extending  [misc]
```

[Pyright](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByYMASgKYCOArkiGQCYA0U51tDzAKogwCJIBjGAChhAgDYBDAM7SoAQQAU3BH0EwAlAC5hUPVElbCxVjTr0A2qhgBdXfoBGR0%2B0vW7YqbKgAhRfO17PUMWSjMGKxRbIKgnY1Iw10jooA) produces no errors.

(And for good measure, the CPython runtime also has a bug: python/cpython#112509.)

I think mypy is right here and both cases should be rejected:

- Overwriting a NotRequired with a Required key is unsound. Given a value `x` of type `A`, it's legal to do `del a["a"]`, but if `x` is of type `B`, that's not allowed.
- Overwriting a Required with a NotRequired doesn't make sense. A value of type `A` must have a key `"b"`, so a type that doesn't require `"b"` cannot be a subtype of `A`.

This logic would be a bit different if PEP-705 is accepted.

**Comment by erictraut** [comment]
I'm not convinced these should be errors. `TypedDict` is used to define a structural type. A `TypedDict` therefore doesn't honor normal class hierarchies.

Protocols are also used to define structural types, but a protocol can define inheritable default implementations. A `TypedDict` does not support inheritable methods (or methods of any kind). Also, in the case of protocols, both pyright and mypy use base classes as a performance shortcut for determining type compatibility between two protocols, but neither of them do this with `TypedDict`. That means there is no type safety issue posed by creating a new `TypedDict` that extends an existing `TypedDict` and redefines fields.

At best, this is an error that's telling the user they may be doing something that they don't intend. That may be appropriate for a linter, but I'm not convinced it's something a type checker should be flagging as an error.

I'll note that pyright correctly reports that a value of type `A` is incompatible with `B` and vice versa when using the two types in the code sample above.

```python
def func(a: A, b: B):
    a = b # Type violation
    b = a # Type violation
```

Pyright is arguably inconsistent in that it emits an error when overriding a `TypedDict` field with an incompatible type, but it allows modifying the "required" attribute of a field. It should probably emit an error in neither of these cases or in both.

```python
class A(TypedDict):
    a: int
    b: int | str
    c: int
    d: ReadOnly[int | str]

class B(A):
    a: str   # Pyright emits error
    b: str   # Pyright emits error
    c: int   # No error
    d: str   # No error
```

Mypy seems to be overly restrictive in that it doesn't allow overriding a `TypedDict` field even if the definition is exactly the same in the extended class definition.

```python
class A(TypedDict):
    a: int

class B(A):
    a: int  # Mypy emits error here even though `a` is declared with exactly the same type
    b: int
```

Here's another interesting datapoint. TypeScript, which uses structural typing exclusive, _does_ complain if you attempt to override a field with an incompatible type. However, it treats fields as covariant, even when they're mutable.

```typescript
interface A1 { a: number }
interface B1 extends A1 { a: string } // Error

interface A2 { a: number | string }
interface B2 extends A2 { a: string } // No error
```

TypeScript doesn't complain if you attempt to redefine a "not required" field as "required", but it does complain in the opposite direction.

```typescript
export interface A3 { a?: number }
export interface B3 extends A3 { a: number } // No error

export interface A4 { a: number }
export interface B4 extends A4 { a?: number } // Error
```

And TypeScript doesn't complain if you attempt to redefine a "readonly" field as writable or vice versa.

```typescript
export interface A5 { readonly a: number }
export interface B5 extends A5 { a: number } // No error

export interface A6 { a: number }
export interface B6 extends A6 { readonly a: number } // No error
```

It looks like TypeScript emits errors only when the resulting type would no longer be type compatible with the type it extends. This makes sense, especially for developers who are used to class hierarchies and nominal typing rules. It also gives the type checker the option of using a performance shortcut when determining type compatibility.

I can see three defensible (type-safe and internally consistent) options:
1. Allow any redeclarations for TypedDict fields.
2. Allow redeclarations of fields as long as the resulting `TypedDict` remains type compatible with the `TypedDict`(s) it extends.
3. Disallow any redeclaration.

Of these, I think option 2 strikes the right balance. This is what TypeScript does. It's also consistent with what draft PEP 705 proposes for the `ReadOnly` attribute.

If we go with option 2, then pyright and mypy would both need to make minor changes.

**Reply by erictraut** [reply]
In my previous post, I forgot about `TypedDict` definitions that extend multiple other `TypedDict` definitions. If we go with option 2 or 3, it should also apply to the multiple inheritance case.

```python
from typing import NotRequired, Required, TypedDict

class A1(TypedDict):
    a: NotRequired[str]

class A2(TypedDict):
    a: Required[str]

class B(A1, A2): # Should generate an error
    ...
```

**Comment by alicederyn** [comment]
> 1. Allow any redeclarations for TypedDict fields.
> 2. Allow redeclarations of fields as long as the resulting TypedDict remains type compatible with the TypedDict(s) it extends.
> 3. Disallow any redeclaration.
> 
> Of these, I think option 2 strikes the right balance. This is what TypeScript does. It's also consistent with what draft PEP 705 proposes for the ReadOnly attribute.

PEP-705 was written under the assumption that a subclass must be structurally compatible with all of its superclasses. That is:

```python
class A(TypedDict):
  foo: Sequence[int]
  bar: ReadOnly[Sequence[int]]
  
class B(A):
  foo: list[int]  # type error
  
class C(A):
  bar: list[int]  # not a type error
```

I think this is option 3 in your list (assuming I understand 2 correctly from "this is what Typescript does"). It would avoid this situation:

```python
b: B = { ... }
a: A = b  # Type error, even though B inherited from A!
```

**Reply by erictraut** [reply]
> I think this is option 3

This is option 2 in my list. Option 3 makes it an error to redefine any field even if the resulting class is still structurally compatible with its superclasses. This would preclude some of the specific use cases you discuss in PEP 705, which says "Subclasses can narrow value types of read-only items". With option 3, this wouldn't be possible.

**Comment by alicederyn** [comment]
I don't think it can be option 2 if option 2 is supposed to include the examples you give of legal TypeScript. For instance, you gave

```typescript
export interface A6 { a: number }
export interface B6 extends A6 { readonly a: number } // No error
```

This would map to

```python
class A6(TypedDict):
  a: int | float

class B6(A6):
  a: ReadOnly[int | float]
```

I was fully intending for this to be rejected by type-checkers, because otherwise we get this:

```python
class C6(B6):
  a: int

c: C6 = { "a": 5 }
b: B6 = c
a: A6 = b  # type error: 'a' is read-only in B6 but not in A6
a['a'] = 3.4  # now c is not a valid C6
```

So perhaps we have *four* options:

1. Allow any redeclarations for TypedDict fields.
2. Allow redeclarations of fields in any way Typescript would allow
3. Allow redeclarations of fields as long as the resulting TypedDict remains structurally compatible with the TypedDict(s) it extends. (Assumption of PEP-705)
4. Disallow any redeclaration.

where 3 and 4 are identical if you do not use PEP-705/ReadOnly.

**Comment by alicederyn** [comment]
I've quickly double-checked the wording of PEP-589 and PEP-705 and as far as I can tell they collectively rule out all but option 3.

1. Ruled out by PEP-589 (["Changing a field type of a parent TypedDict class in a subclass is not allowed."](https://peps.python.org/pep-0589/#inheritance:~:text=Changing%20a%20field%20type%20of%20a%20parent%20TypedDict%20class%20in%20a%20subclass%20is%20not%20allowed.))
2. Ruled out by PEP-589 (same sentence)
3. Compatible with both
4. Ruled out by PEP-705 (["Subclasses can narrow value types of read-only items"](https://peps.python.org/pep-0705/#inheritance:~:text=Subclasses%20can%20narrow%20value%20types%20of%20read%2Donly%20items%3A))

The TypeScript example I'm using to conclude PEP-589 rules out 2 is:

```typescript
interface A2 { a: number | string }
interface B2 extends A2 { a: string } // No error
```

That said:

* the text of PEP-705 only rules out 4 because I assumed inheritance was supposed to match structural subtyping, and wanted to make it clear what that entailed. I will happily change it to reflect any decision here.
* there are other possible rulesets compatible with both PEP-589 and PEP-705 that are weaker than option 3, depending on whether you interpret "field type" to include required/not-required or not; they just won't match TypeScript

As long as all type-checkers agree on what the actual _meaning_ is for anything they mutually admit as legal, I like Eric's statement:

> [This] may be appropriate for a linter

**Reply by erictraut** [reply]
Sorry if I confused folks by bringing up TypeScript. I wasn't suggesting that we should adopt TypeScript's rules for structural type compatibility. I'm suggesting that we should stick with Python's existing rules for structural type compatibility and "allow redeclaration of TypedDict fields as long as the resulting TypedDict remains structurally compatible with the TypedDict it extends". That's the "option 2" in my list (and "option 3" in your revised list above). It sounds like we're in agreement.

> ...3 and 4 are identical if you do not use PEP-705/ReadOnly

Almost, but not quite. The following is allowed with the former but not the latter:

```python
class A(TypedDict):
    x: Required[int]

class B(A):
    x: Required[int]
```

Mypy disallows this currently, but it would be allowed under (my) "option 2" today prior to the introduction of `ReadOnly`. I admit that it's an edge case today, but it's important to get the principle correct for future extensions like `ReadOnly`.

**Reply by erictraut** [reply]
I've logged two issues in the pyright issue tracker:
* https://github.com/microsoft/pyright/issues/6588
* https://github.com/microsoft/pyright/issues/6589

**Reply by JelleZijlstra** [reply]
Here's another case where 3 and 4 may diverge without PEP 705:

```python
class A(TypedDict):
    x: Any
class B(A):
    x: int
```

Pyright [currently complains](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFACqICmAJgCJIDGMAUIwDYCGAzt4QBSUEtBswCUALhZQZUAB4TCxNl15QAQvwKTpshZhSsgA) about this, but I think it shouldn't.

**Reply by erictraut** [reply]
> Pyright currently complains about this, but I think it shouldn't.

I agree. Just logged [this issue](https://github.com/microsoft/pyright/issues/6590).

**Comment by alicederyn** [comment]
Do we want this to apply to Protocols too? I have come across the following interesting case:

```python
class A(Protocol):
  x: Final[int | None]

class B(Protocol):
  x: Final[int]

def b_to_a(b: B) -> A:
  return b

class C(A, Protocol):
  x: Final[int]
```

Passing this into mypy, I get:

```text
error: Protocol member cannot be final
error: Final name must be initialized with a value
...
```

I'm not sure why mypy is preventing use of Final on Protocols. (Maybe to avoid this whole debate!)

On the other hand, pyright gives me just:

```text
error: "x" cannot be redeclared because parent class "A" declares it as Final
```

But no error on the line `return b`. So pyright considers B a valid structural subtype of A, but won't let this be enforced at declaration time with an explicit use of subclassing. This would be more obvious to me if A was also providing an implementation of that final value, but unlike mypy, pyright isn't requiring that.

Is this because it _could_ provide a value, and pyright doesn't track that, so has to assume it _does_? Or is this just stemming from a literal reading of "prevents redefinition in subclasses" in the PEP?

(This came up because someone asked for clarification on why I didn't reuse `Final` in PEP-705, and I realized this was another place I was making assumptions about typechecking behavior based on my prior that subclassing a structural type is just type-checked subtyping.)
